Framework/Algorithms/src/CreateMonteCarloWorkspace.cpp

```diff
@@ -0,0 +1,169 @@
+}
+
+//----------------------------------------------------------------------------------------------
+
+Mantid::HistogramData::HistogramY CreateMonteCarloWorkspace::fillHistogramWithRandomData(const std::vector<double> &cdf,
+                                                                                         int numIterations,
+                                                                                         int seed_input,
+                                                                                         API::Progress &progress) {
+
+  Mantid::HistogramData::HistogramY outputY(cdf.size(), 0.0);
+  std::mt19937 gen(seed_input);
+  std::uniform_real_distribution<> dis(0.0, 1.0);
+
+  int progressInterval = std::max(1, numIterations / 100); // Update progress every 1%
+
+  for (int i = 0; i < numIterations; ++i) {
+    double randomNum = dis(gen);
+    auto it = std::lower_bound(cdf.begin(), cdf.end(), randomNum);
+    size_t index = std::distance(cdf.begin(), it);
+
+    if (index < outputY.size()) {
+      outputY[index] += 1.0;
+    }
+
+    if (i % progressInterval == 0) {
+      progress.report("Generating random data...");
+    }
+  }
+  return outputY;
+}
+
+/**
+ *  Compute a normalized CDF [0..1] from the given histogram data.
+ */
+std::vector<double> CreateMonteCarloWorkspace::computeNormalizedCDF(const Mantid::HistogramData::HistogramY &yData) {
+  std::vector<double> cdf(yData.size());
+  std::partial_sum(yData.begin(), yData.end(), cdf.begin());
+  double total_counts = cdf.back();
+
+  if (total_counts > 0.0) {
+    // Normalize the CDF
+    std::transform(cdf.begin(), cdf.end(), cdf.begin(), [total_counts](double val) { return val / total_counts; });
+  } else {
+    g_log.warning("Total counts are zero; normalization skipped.");
+  }
+  return cdf;
+}
+
+/**
+ *  Determine how many iterations to use for MC sampling.
+ *  If userMCEvents > 0, use that directly; otherwise use the integral of the input data.
+ */
+int CreateMonteCarloWorkspace::computeNumberOfIterations(const Mantid::HistogramData::HistogramY &yData,
+                                                         int userMCEvents) {
+  if (userMCEvents > 0) {
+    return userMCEvents;
+  }
+
+  // Default: Integral of Input
+  double total_counts = std::accumulate(yData.begin(), yData.end(), 0.0);
+  int iterations = static_cast<int>(std::round(total_counts));
+
+  if (iterations == 0) {
+    g_log.warning("Total counts in the input workspace round to 0. No Monte Carlo events will be generated.");
+  }
+  return iterations;
+}
+
+//----------------------------------------------------------------------------------------------
+/** Execute the algorithm.
+ */
+void CreateMonteCarloWorkspace::exec() {
+  MatrixWorkspace_sptr inputWs = getProperty("InputWorkspace");
+  int seed_input = getProperty("Seed");
+  int userMCEvents = getProperty("MonteCarloEvents");
+
+  const Mantid::HistogramData::HistogramY &originalYData = inputWs->y(0); // Counts in each bin
```

Comment:
Think it's preferred to use `auto` where possible, here's one example where I think you could
```suggestion
  const auto &originalYData = inputWs->y(0); // Counts in each bin
```